server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreChoosingModule.java

```diff
@@ -43,11 +49,47 @@
 import com.google.common.annotations.VisibleForTesting;
 import com.google.inject.AbstractModule;
 import com.google.inject.Provides;
+import com.google.inject.Scopes;
 import com.google.inject.multibindings.Multibinder;
+import com.google.inject.name.Names;
 
 public class BlobStoreChoosingModule extends AbstractModule {
     private static final Logger LOGGER = LoggerFactory.getLogger(BlobStoreChoosingModule.class);
 
+    public static class CacheDisabledModule extends AbstractModule  {
```

Comment:
why it's so complicated? Can't we just bind the right implementation of BlobStore in `BlobStoreChoosingModule.configure`?

Reply:
No not exactly. 

Let me recall you the `cahier des charges`: https://github.com/linagora/james-project/pull/3261#pullrequestreview-389804841

```
If I understand correctly, whether you use the cassandra cache feature or not, you are going to create a connection for it?

Please make sure we can use Cassandra without the cache feature without paying the cost.
```

Here we are **sure** we are not initializing the session if it is not configured. Every if based, provider based technique, I cannot trust it. Sorry.

And **especially** https://github.com/linagora/james-project/pull/3261#issuecomment-613349211

```
We expect James to add features that are plugable.
In this case, Cassandra blob cache makes sense only when using ObjectStorage and cache is enabled.
It should not be injected into james into every other cases.
```

Here you have it....

Reply:
>  Every if based, provider based technique, I cannot trust it. Sorry.

Could you elaborate?

To me:

```
if (cached) {
  bind(BlobStore.class).annotatedWith("backend").to(Foo.class);
  bind(BlobStore.class).to(CacheBlobStore.class);
} else {
  bind(BlobStore.class).to(Foo.class);
}
```

seems OK.

Reply:
CassandraCacheModule would be injected anyway. I fear it pulls dependency to the Cassandra Cache session.

Also, given the list of modules installed, I can have two very distinct James servers (in their behaviours)

IMO the guice module composition should reflect the architecture of the James server we rely on. Conditional logic in Modules should thus be avoided as it defeat that design goal.

Reply:
```
> Should we have the DistributedJamesServerMain doing a lookup on the configuration to know if it should compose the cache module?

yes
```

<3

https://github.com/linagora/james-project/pull/3261#issuecomment-613911695

(14 days ago)